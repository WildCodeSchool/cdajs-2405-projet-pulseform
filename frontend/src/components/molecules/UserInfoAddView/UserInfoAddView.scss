@use "@assets/variables" as *;

.user-info-add-view {
  height: 100vh;

  &__container {
    &__form {
      display: flex;
      flex-direction: column;
      align-items: stretch;
      width: 16rem;
      gap: 1rem;

      &__container {
        display: flex;
        align-items: stretch;
        justify-content: space-between;

        &__value-name {
          @include small-phone;
          display: flex;
          align-items: center;
          justify-content: center;
          background-color: $tertiary;
          color: $white;
          width: 2.5rem;
          height: 2.5rem;
          border-radius: 10px;
        }
      }

      &__btn {
        width: 100%;
      }
    }
  }
}

.user-info-add-view__container__form__btn--transparent {
  background: transparent;
  border: none;
  color: $white;
  text-decoration: underline;
  font-size: 1rem;
  align-self: center;

  &:hover {
    background: transparent;
    cursor: pointer;
  }
}
