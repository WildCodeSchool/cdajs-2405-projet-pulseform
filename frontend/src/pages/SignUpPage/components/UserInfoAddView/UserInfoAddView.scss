@import "@assets/variables.scss";

.user-info-add-view {
  height: 100vh;

  &__container {
    &__form {
      display: flex;
      flex-direction: column;
      align-items: stretch;
      width: 16rem;
      gap: 1rem;

      &__container {
        display: flex;
        align-items: stretch;
        justify-content: space-between;

        &__value-name {
          @include small-phone;
          display: flex;
          align-items: center;
          justify-content: center;
          background-color: $tertiary;
          color: $white;
          width: 2.5rem;
          height: 2.5rem;
          border-radius: 10px;
        }
      }

      &__btn {
        width: 100%;
      }
    }
  }
}
