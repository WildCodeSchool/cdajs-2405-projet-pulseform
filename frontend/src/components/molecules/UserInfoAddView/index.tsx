import { useMutation } from "@apollo/client";
import { useState } from "react";
import { useTranslation } from "react-i18next";

import BasicButton from "@components/atoms/BasicButton";
import BodyStepQuestions from "@components/molecules/BodyStepQuestions";
import CompleteLaterButton from "@components/molecules/CompleteLaterButton/CompleteLaterButton";
import ExitSignUpStep from "@components/molecules/ExitSignUpStep";
import GenderSelectView from "@components/molecules/GenderSelectView";

import { useUser } from "@context/UserContext";
import { FitnessLevel, MemberRole } from "@graphql/__generated__/schema";
import { UPDATE_USER_MUTATION } from "@graphql/mutations/user";

import "./UserInfoAddView.scss";

function UserInfoAddView() {
  const { t } = useTranslation();
  const { user } = useUser();

  const [birthday, setBirthday] = useState("");
  const [weight, setWeight] = useState("");
  const [height, setHeight] = useState("");
  const [isExit, setIsExit] = useState(false);
  const [isNextStep, setIsNextStep] = useState(false);
  const [formError, setFormError] = useState("");

  const [updateUser] = useMutation(UPDATE_USER_MUTATION, {
    refetchQueries: ["Me"],
  });

  const handleExit = () => setIsExit(true);
  const handleNextStep = () => setIsNextStep(true);

  const onSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    setFormError("");

    if (!birthday || !weight || !height) {
      setFormError("Veuillez remplir tous les champs.");
      return;
    }

    const birthdayDate = new Date(birthday);
    if (Number.isNaN(birthdayDate.getTime())) {
      setFormError("Date de naissance invalide.");
      return;
    }

    if (!user) return;

    const today = new Date();
    const month = today.toLocaleString("en-US", { month: "short" });

    try {
      await updateUser({
        variables: {
          data: {
            id: Number(user.id),
            username: user.username,
            email: user.email,
            password: "",
            description: user.description ?? "",
            image: user.image ?? "",
            role: user.role ?? MemberRole.User,
            created_at: new Date(user.created_at),
            level: user.level ?? FitnessLevel.Beginner,
            birthday: birthdayDate,
            height: Number(height),
            weights: [
              {
                month,
                weight: Number(weight),
              },
            ],
          },
        },
      });

      handleNextStep();
    } catch (err) {
      console.error("‚ùå Erreur lors de l'envoi :", err);
      setFormError("Une erreur est survenue.");
    }
  };

  return (
    <>
      {!isExit && !isNextStep && (
        <section className="user-info-add-view">
          <BodyStepQuestions
            ctaExit={handleExit}
            questionLabel={t("MORE_ABOUT_YOU")}
          >
            <form
              className="user-info-add-view__container__form"
              onSubmit={onSubmit}
            >
              <input
                type="date"
                value={birthday}
                onChange={(e) => setBirthday(e.target.value)}
                placeholder="JJ/MM/AAAA"
              />

              <div className="user-info-add-view__container__form__container">
                <input
                  type="number"
                  value={weight}
                  onChange={(e) => setWeight(e.target.value)}
                  placeholder="Poids"
                />
                <p className="user-info-add-view__container__form__container__value-name">
                  KG
                </p>
              </div>

              <div className="user-info-add-view__container__form__container">
                <input
                  type="number"
                  value={height}
                  onChange={(e) => setHeight(e.target.value)}
                  placeholder="Taille"
                />
                <p className="user-info-add-view__container__form__container__value-name">
                  CM
                </p>
              </div>

              {formError && (
                <p style={{ color: "red", fontSize: "0.8rem" }}>{formError}</p>
              )}

              <BasicButton
                className="user-info-add-view__container__form__btn"
                type="submit"
              >
                {t("NEXT")}
              </BasicButton>

              <CompleteLaterButton />
            </form>
          </BodyStepQuestions>
        </section>
      )}
      {isExit && !isNextStep && <ExitSignUpStep />}
      {!isExit && isNextStep && <GenderSelectView />}
    </>
  );
}

export default UserInfoAddView;
