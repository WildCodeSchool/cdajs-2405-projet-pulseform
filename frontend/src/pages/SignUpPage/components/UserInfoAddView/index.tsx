import { useState } from "react";
import { useTranslation } from "react-i18next";

import BasicButton from "@components/atoms/BasicButton";
import BodyStepQuestions from "@components/molecules/BodyStepQuestions";
import ExitSignUpStep from "../ExitSignUpStep";
import GenderSelectView from "../GenderSelectView";

import "./UserInfoAddView.scss";

function UserInfoAddView() {
  const [isExit, setIsExit] = useState(false);
  const [isNextStep, setIsNextStep] = useState(false);

  const { t } = useTranslation();

  function handleExit() {
    setIsExit(true);
  }

  function handleNextStep() {
    setIsNextStep(true);
  }

  return (
    <>
      {!isExit && !isNextStep && (
        <section className="user-info-add-view">
          <BodyStepQuestions
            ctaExit={handleExit}
            questionLabel={t("MORE_ABOUT_YOU")}
          >
            <form className="user-info-add-view__container__form" action="">
              <input type="date" />
              <div className="user-info-add-view__container__form__container">
                <input type="number" />
                <p className="user-info-add-view__container__form__container__value-name">
                  KG
                </p>
              </div>
              <div className="user-info-add-view__container__form__container">
                <input type="number" />
                <p className="user-info-add-view__container__form__container__value-name">
                  CM
                </p>
              </div>
              <BasicButton
                className="user-info-add-view__container__form__btn"
                type="submit"
                onClick={handleNextStep}
              >
                {t("NEXT")}
              </BasicButton>
            </form>
          </BodyStepQuestions>
        </section>
      )}
      {isExit && !isNextStep && <ExitSignUpStep />}
      {!isExit && isNextStep && <GenderSelectView />}
    </>
  );
}

export default UserInfoAddView;
