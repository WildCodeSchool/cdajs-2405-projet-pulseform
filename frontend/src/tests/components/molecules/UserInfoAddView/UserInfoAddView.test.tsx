import UserInfoAddView from "@components/molecules/UserInfoAddView";
import { render } from "@testing-library/react";

describe("UserInfoAddView", () => {
  it("renders the UserInfoAddView component", () => {
    const tree = render(<UserInfoAddView />);
    expect(tree).toMatchSnapshot();
  });
});
