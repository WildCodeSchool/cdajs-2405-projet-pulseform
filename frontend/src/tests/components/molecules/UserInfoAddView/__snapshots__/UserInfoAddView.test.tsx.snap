// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserInfoAddView > renders the UserInfoAddView component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="user-info-add-view"
      >
        <div
          class="body-step-questions"
        >
          <div
            class="body-step-questions__header-d"
          >
            <div
              class="body-step-questions__header-d__logo"
            >
              <div
                class="little-logo little-logo--desktop"
              >
                <img
                  alt="Logo Pulse Form"
                  class="little-logo__img"
                  src="/src/assets/icons/mobile-logo.svg"
                />
                <p
                  class="little-logo__title little-logo__title--desktop"
                >
                  Pulse Form
                </p>
              </div>
            </div>
            <div
              class="body-step-questions__header-d__container"
            >
              <button
                class="body-step-questions__header-d__container__btn"
                type="button"
              >
                TEST_PROGRAM
              </button>
            </div>
          </div>
          <button
            class="body-step-questions__exit"
            type="button"
          >
            <img
              alt="Close"
              src="/src/assets/icons/blue-cross.svg"
            />
          </button>
          <h2
            class="body-step-questions__question"
          >
            MORE_ABOUT_YOU
          </h2>
          <img
            alt="blob-girl"
            class="body-step-questions__blob-girl"
            src="/src/assets/images/blob-girl.svg"
          />
          <div
            class="body-step-questions__container"
          >
            <form
              action=""
              class="user-info-add-view__container__form"
            >
              <input
                type="date"
              />
              <div
                class="user-info-add-view__container__form__container"
              >
                <input
                  type="number"
                />
                <p
                  class="user-info-add-view__container__form__container__value-name"
                >
                  KG
                </p>
              </div>
              <div
                class="user-info-add-view__container__form__container"
              >
                <input
                  type="number"
                />
                <p
                  class="user-info-add-view__container__form__container__value-name"
                >
                  CM
                </p>
              </div>
              <button
                class="basic-button user-info-add-view__container__form__btn basic-button--basic"
                type="submit"
              >
                NEXT
              </button>
            </form>
          </div>
          <div
            class="body-step-questions__container-d"
          >
            <div
              class="body-step-questions__container-d__question"
            >
              <img
                alt="sport"
                class="body-step-questions__container-d__question__img"
                src="/src/assets/images/step-img.svg"
              />
              <div
                class="body-step-questions__container-d__question__container"
              >
                <p
                  class="body-step-questions__container-d__question__container__label"
                >
                  MOTIVATIONAL_LABEL
                </p>
              </div>
            </div>
            <div
              class="body-step-questions__container-d__children"
            >
              <p
                class="body-step-questions__container-d__children__label"
              >
                MORE_ABOUT_YOU
              </p>
              <form
                action=""
                class="user-info-add-view__container__form"
              >
                <input
                  type="date"
                />
                <div
                  class="user-info-add-view__container__form__container"
                >
                  <input
                    type="number"
                  />
                  <p
                    class="user-info-add-view__container__form__container__value-name"
                  >
                    KG
                  </p>
                </div>
                <div
                  class="user-info-add-view__container__form__container"
                >
                  <input
                    type="number"
                  />
                  <p
                    class="user-info-add-view__container__form__container__value-name"
                  >
                    CM
                  </p>
                </div>
                <button
                  class="basic-button user-info-add-view__container__form__btn basic-button--basic"
                  type="submit"
                >
                  NEXT
                </button>
              </form>
            </div>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="user-info-add-view"
    >
      <div
        class="body-step-questions"
      >
        <div
          class="body-step-questions__header-d"
        >
          <div
            class="body-step-questions__header-d__logo"
          >
            <div
              class="little-logo little-logo--desktop"
            >
              <img
                alt="Logo Pulse Form"
                class="little-logo__img"
                src="/src/assets/icons/mobile-logo.svg"
              />
              <p
                class="little-logo__title little-logo__title--desktop"
              >
                Pulse Form
              </p>
            </div>
          </div>
          <div
            class="body-step-questions__header-d__container"
          >
            <button
              class="body-step-questions__header-d__container__btn"
              type="button"
            >
              TEST_PROGRAM
            </button>
          </div>
        </div>
        <button
          class="body-step-questions__exit"
          type="button"
        >
          <img
            alt="Close"
            src="/src/assets/icons/blue-cross.svg"
          />
        </button>
        <h2
          class="body-step-questions__question"
        >
          MORE_ABOUT_YOU
        </h2>
        <img
          alt="blob-girl"
          class="body-step-questions__blob-girl"
          src="/src/assets/images/blob-girl.svg"
        />
        <div
          class="body-step-questions__container"
        >
          <form
            action=""
            class="user-info-add-view__container__form"
          >
            <input
              type="date"
            />
            <div
              class="user-info-add-view__container__form__container"
            >
              <input
                type="number"
              />
              <p
                class="user-info-add-view__container__form__container__value-name"
              >
                KG
              </p>
            </div>
            <div
              class="user-info-add-view__container__form__container"
            >
              <input
                type="number"
              />
              <p
                class="user-info-add-view__container__form__container__value-name"
              >
                CM
              </p>
            </div>
            <button
              class="basic-button user-info-add-view__container__form__btn basic-button--basic"
              type="submit"
            >
              NEXT
            </button>
          </form>
        </div>
        <div
          class="body-step-questions__container-d"
        >
          <div
            class="body-step-questions__container-d__question"
          >
            <img
              alt="sport"
              class="body-step-questions__container-d__question__img"
              src="/src/assets/images/step-img.svg"
            />
            <div
              class="body-step-questions__container-d__question__container"
            >
              <p
                class="body-step-questions__container-d__question__container__label"
              >
                MOTIVATIONAL_LABEL
              </p>
            </div>
          </div>
          <div
            class="body-step-questions__container-d__children"
          >
            <p
              class="body-step-questions__container-d__children__label"
            >
              MORE_ABOUT_YOU
            </p>
            <form
              action=""
              class="user-info-add-view__container__form"
            >
              <input
                type="date"
              />
              <div
                class="user-info-add-view__container__form__container"
              >
                <input
                  type="number"
                />
                <p
                  class="user-info-add-view__container__form__container__value-name"
                >
                  KG
                </p>
              </div>
              <div
                class="user-info-add-view__container__form__container"
              >
                <input
                  type="number"
                />
                <p
                  class="user-info-add-view__container__form__container__value-name"
                >
                  CM
                </p>
              </div>
              <button
                class="basic-button user-info-add-view__container__form__btn basic-button--basic"
                type="submit"
              >
                NEXT
              </button>
            </form>
          </div>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
